File: research lab/processing.py
import qutip as qt
from qutip.core.gates import rx, ry, hadamard_transform
import numpy as np
import jax.numpy as jnp
import jax

from itertools import chain 
from jax import random
import optax
from tqdm.auto import tqdm
import time
from numpy.random import default_rng
from itertools import product, chain


def cnot_gate(n, control, target):

    """
    Create a CNOT gate for an n-qubit system with the specified control and target qubits.

    Parameters:
        n (int): Number of qubits in the system.
        control (int): Index of the control qubit (0-based).
        target (int): Index of the target qubit (0-based).

    Returns:
        Qobj: The CNOT gate for the n-qubit system.
    """
    # Identity operator for a single qubit
    I = qt.qeye(2)
    # Pauli-X gate for a single qubit
    X = qt.sigmax()

    # Create the projection operators for the control qubit
    P0 = qt.basis(2, 0) * qt.basis(2, 0).dag()  # |0><0|
    P1 = qt.basis(2, 1) * qt.basis(2, 1).dag()  # |1><1|

    # Start building the full CNOT gate
    # Part 1: When the control qubit is |0>, no change
    CNOT = qt.tensor(
        *[
            I if i != control else P0
            for i in range(n)
        ]
    ) 
    # Part 2: When the control qubit is |1>, apply X to the target qubit 
    CNOT += qt.tensor(
        *[
            I if i != control and i != target
            else (P1 if i == control else X)
            for i in range(n)
        ]
    )

    return CNOT

def parse_circuit_to_qobj(text_circuits, n):
    """
    Converts a list of text-based quantum circuits into Qutip unitary operators.

    Args:
        text_circuits (list of str): List of text circuit descriptions.
        n (int): Number of qubits.

    Returns:
        list of Qobj: List of Qutip unitary operators of dimension [[2]^n, [2]^n].
    """
    unitary_list = []

    for circuit in text_circuits:
        U = qt.qeye([2] * n)  # Start with identity matrix
        
        # Extract operations from the text
        operations = circuit.split(")(")  
        operations = [op.strip("()") for op in operations if op.strip("()")]  # Remove empty entries

        for op in operations:
            gate_info = op.split(":")
            if len(gate_info) < 2:
                continue  # Skip invalid formats

            gate_name = gate_info[0]
            qubit_indices = list(map(int, gate_info[1].split(",")))

            # Apply the correct gate
            if gate_name == "RX90":
                U = qt.tensor([rx(np.pi/2) if i == qubit_indices[0] else qt.qeye(2) for i in range(n)]) * U
            elif gate_name == "RY90":
                U = qt.tensor([ry(np.pi/2) if i == qubit_indices[0] else qt.qeye(2) for i in range(n)]) * U
            elif gate_name == "H":
                U = qt.tensor([hadamard_transform(N=1) if i == qubit_indices[0] else qt.qeye(2) for i in range(n)]) * U
            elif gate_name == "CNOT":
                U = cnot_gate(n, qubit_indices[0], qubit_indices[1]) * U
            else:
                raise ValueError(f"Unsupported gate: {gate_name}")

        unitary_list.append(U)

    return unitary_list


# Prepare State
def prepare_state(initial_text,N):
    U_state_prep=parse_circuit_to_qobj([initial_text], N)
    initial_state= qt.tensor([qt.basis(2,0)]*N)
    rho_ideal=U_state_prep[0]*initial_state
    rho_ideal=rho_ideal*rho_ideal.dag()
    return rho_ideal

def flatten_list(nested_list):
    """Flattens a list of lists into a single list using itertools.chain."""
    return list(chain(*nested_list))


def block_seperator(rho , selective_blocks, N):
    """
    Extracts and sums specific block structures from a given density matrix based on selective binary representations.
    
    Args:
        rho (numpy.ndarray): Input density matrix of shape (2^N, 2^N).
        selective_blocks (list of int): List of integers representing selective blocks to extract (must be < 2^N).
        N (int): Number of qubits in the system.

    Returns:
        qutip.Qobj: Resulting matrix as a Qobj with dimension (2^N, 2^N), representing the sum of selected blocks.
    """
    
    assert all(0 <= block < 2**N for block in selective_blocks), "All entries in selective_blocks must be less than 2**N"
    
    binary_reps = [format(block, f'0{N}b') for block in selective_blocks]
    
    pauli_observables = []
    for binary in binary_reps:
        operators = [qt.qeye(2) if bit == '0' else qt.sigmax() for bit in binary]
        pauli_observables.append(qt.tensor(*operators))
    
    total_observable = sum(obs.full() * rho for obs in pauli_observables)
    
    return qt.Qobj(total_observable, dims=[[2]*N, [2]*N])




@jax.jit
def data_predict_from_rho(rho, circuits, shots):
    """
    Predict measurement outcomes from a given quantum state (rho) and a set of quantum circuits.
    
    Args:
        rho (jnp.array): Density matrix representing the quantum state.
        circuits (jnp.array): Array of quantum circuit matrices applied to the state.
        shots (int ): Number of measurement shots to scale probabilities.

    Returns:
        jnp.array: Predicted measurement probabilities scaled by shots for each circuit shape(circuit,comp basis).
    """
    # Compute predicted measurement probabilities for each circuit
    def compute_for_circuit(circuit):
        return jnp.diag(jnp.dot(jnp.dot(circuit, rho), circuit.T.conj())) * shots
    
    # Use vmap to apply computation across all circuits efficiently
    data_predict = jax.vmap(compute_for_circuit)(circuits)
    
    return jnp.real(data_predict)

def data_predict_from_rho_sampled(rho, circuits, shots):
    """
    Predict measurement outcomes from a quantum state (rho) and circuits,
    sampling finite-shot counts from the probability distribution.

    Args:
        rho (jnp.array): Density matrix.
        circuits (jnp.array): Array of circuit unitaries.
        shots (int): Number of measurement shots.

    Returns:
        jnp.array: Sampled counts (integer array) for each circuit.
    """

    def compute_probs(circuit):
        probs = jnp.real(jnp.diag(circuit @ rho @ circuit.T.conj()))
        probs = probs / jnp.sum(probs)  # normalize to sum to 1
        return probs

    # Step 1: compute all probabilities with vmap
    probs_all = jax.vmap(compute_probs)(circuits)

    # Step 2: sample counts with NumPy loop
    counts_list = []
    for probs in np.array(probs_all):  # convert to numpy array
        counts = np.random.multinomial(shots, probs)
        counts_list.append(counts)

    return jnp.array(counts_list)


def random_matrix_jax(N, key=random.PRNGKey(0)):
    """
    Generate a random (non-Hermitian) matrix using JAX for an N-qubit system.
    
    Parameters:
        N (int): Number of qubits.
        key: JAX PRNG key.
        
    Returns:
        jnp.array: random complex matrix
    """
    # Total Hilbert space dimension for N qubits
    dim = 2**N
    
    # Generate random real and imaginary parts using JAX
    key_real, key_imag = random.split(key)
    real_part = random.uniform(key_real, (dim, dim))
    imag_part = random.uniform(key_imag, (dim, dim)) * 1j
    
    # Combine real and imaginary parts
    random_matrix = real_part + imag_part
    
    
    return random_matrix


# Compute the density matrix rho = A^T A / tr(A^T A)
@jax.jit
def density_matrix(A):
    AtA = jnp.dot(A.T.conj(), A)
    rho = AtA / jnp.trace(AtA)
    return rho 

# Define the log loss function (user-defined)
@jax.jit
def log_loss_function(A,data,circuits,shots):
    rho = density_matrix(A)
    predicted_data = data_predict_from_rho(rho,circuits,shots=1)
    return -jnp.sum(data * jnp.log(predicted_data+1e-8))/shots

def gd_chol_rank(data, rho_or, ops_jnp,shots, params: optax.Params, iterations: int,  batch_size: int,
            lr=2e-1, decay = 0.999, lamb:float =0.00001, batch=False, tqdm_off=False):
  """
  Function to do the GD-Chol.
  Return:
    params1: The reconstructed density matrix
    fidelities_GD: A list with the fidelities values per iteration
    timel_GD: A list with the value of the time per iteration
    loss1: A list with the value of the loss function per iteration

  Input:
    data: the expected value of the original density matrix
    rho_or: original density matrix, to calculate the fidelity
    ops_jnp: POVM in jnp array
    params: Ansatz, any complex matrix T (not necessary the lower triangular)
    iterations: number of iterations for the method
    batch_size: batch size
    lr: learning rate
    decay: value of the decay of the lr (defined in given optimizer)
    lamb: hyperparameter l1 regularization
    batch: True to have mini batches, False to take all the data
    tqdm_off: To show the iteration bar. True is to desactivate (for the cluster)
    
  """
  convergence_tol = 1e-10  # minimal change considered as progress
  patience = 50           # number of allowed stagnant steps before stopping
  no_improve_counter = 0
  
  start_learning_rate = lr
  # Exponential decay of the learning rate.
  scheduler = optax.exponential_decay(
      init_value=start_learning_rate, 
      transition_steps=iterations,
      decay_rate=decay)
  # Combining gradient transforms using `optax.chain`.
  gradient_transform = optax.chain(
      optax.clip_by_global_norm(1.0),  # Clip by the gradient by the global norm.
      optax.scale_by_adam(),  # Use the updates from adam.
      optax.scale_by_schedule(scheduler),  # Use the learning rate from the scheduler.
      # Scale updates by -1 since optax.apply_updates is additive and we want to descend on the loss.
      optax.scale(-1.0)
  )
  

  loss1 = []
  fidelities_GD = []
  timel_GD = []
  #par_o = jnp.matmul(jnp.conj(params.T),params)/jnp.trace(jnp.matmul(jnp.conj(params.T),params))
  #fidelities_GD.append(qtp.fidelity(rho_or, qtp.Qobj(par_o)))
  #loss1.append(float(cost(params, jnp.asarray(data), ops_jnp, lamb)))
  opt_state = gradient_transform.init(params)
  num_me = len(data)
  # opt_state = optimizer.init(params)
  if not tqdm_off:
    pbar_GD = tqdm(range(iterations)) 
  
  @jax.jit
  def step(params, opt_state, data, ops_jnp):
    grad_f = jax.grad(log_loss_function, argnums=0)(params,data,ops_jnp,shots)
    grads = jnp.conj(grad_f)           # do a conjugate, if not can create some problems
    # updates, opt_state = optimizer.update(grads, opt_state, params)
    updates, opt_state = gradient_transform.update(grads, opt_state, params)
    params = optax.apply_updates(params, updates)

    return params, opt_state
  

  tot_time = 0
  for i in tqdm(range(iterations), disable=tqdm_off):
    start = time.time()
    if batch:
        rng = default_rng()
        indix = rng.choice(num_me, size=batch_size, replace=False)
        # indix = np.random.randint(0, num_me, size=[batch_size])
        data_b = jnp.asarray(data[[indix]].flatten())
        ops2 = ops_jnp[indix]
    else: 
        ops2 = ops_jnp
        data_b = data
    params, opt_state = step(params, opt_state,data_b, ops2)
    #params = rho_cons(params)
    end = time.time()
    timestep = end - start
    tot_time += timestep
    timel_GD.append(tot_time)    
    
    par1 = jnp.matmul(jnp.conj(params.T),params)/jnp.trace(jnp.matmul(jnp.conj(params.T),params))
    loss1.append(float(log_loss_function(params,data,ops_jnp,shots)))
    f = qt.fidelity(rho_or, qt.Qobj(np.array(par1)))
    fidelities_GD.append(f)
    

    #timel_GD.append(end - start)  
    if not tqdm_off:
        pbar_GD.set_description("Fidelity GD-chol-rank {:.4f}".format(f))
        pbar_GD.update() 
        
    if len(loss1) > 10:
        if abs(loss1[-1] - loss1[-2]) < convergence_tol:
            no_improve_counter += 1
        else:
            no_improve_counter = 0
    
    if no_improve_counter >= patience:
        #print(f"Early stopping at iteration {i} due to convergence.")
        break

  params1 = jnp.matmul(jnp.conj(params.T),params)/jnp.trace(jnp.matmul(jnp.conj(params.T),params))
  return params1, fidelities_GD, timel_GD, loss1



def process_data(data, unitaries_jnp, selective_blocks, shots, N, rho_ideal=None):
    A = random_matrix_jax(N)
    
    # ← DEBUG (ohne L0)
    print("=== DEBUG Python ===")
    print("data shape:         ", data.shape)
    print("data sum:           ", float(data.sum()))
    print("data min:           ", float(data.min()))
    print("data max:           ", float(data.max()))
    print("Anzahl Unitaries:   ", len(unitaries_jnp))
    print("Anzahl Datenpunkte: ", data.size)

    # rest bleibt gleich
    iterations = 100*N*3
    if rho_ideal != None:
        rho_ideal = qt.Qobj(np.array(rho_ideal.full()))
    else:
        rho_ideal = qt.rand_dm(2**N)
    rho, fidelities, time_GD, loss = gd_chol_rank(
        data, rho_ideal, unitaries_jnp, shots, A, iterations,
        100, lr=0.1, decay=0.999, lamb=0.00001, batch=False, tqdm_off=True
    )
    return block_seperator(np.array(rho), selective_blocks, N)

import numpy as np
import jax.numpy as jnp
import qutip as qt
from processing import (
    cnot_gate, parse_circuit_to_qobj, prepare_state,
    flatten_list, block_seperator,
    data_predict_from_rho, data_predict_from_rho_sampled,
    random_matrix_jax, process_data
)

print("═══ Tests Python processing.py ═══\n")

from setup import build_parallel_entangler_blocks
from processing import parse_circuit_to_qobj, data_predict_from_rho_sampled, process_data, flatten_list
import qutip as qt
import jax.numpy as jnp
import numpy as np

N = 4

# Schaltkreise automatisch generieren
blocks    = list(range(2**N))
ent_circs = build_parallel_entangler_blocks(blocks, N)
circuits  = flatten_list(ent_circs)

rho_true = qt.rand_dm(2**N)
Us_all   = jnp.array([u.full() for u in parse_circuit_to_qobj(circuits, N)])
shots    = 1000 * 2**N
data     = data_predict_from_rho_sampled(jnp.array(rho_true.full()), Us_all, shots)

rho_rec  = process_data(
    data             = data,
    unitaries_jnp    = Us_all,
    selective_blocks = blocks,
    shots            = shots,
    N                = N,
    rho_ideal        = rho_true
)

rho_rec_qobj  = qt.Qobj(np.array(rho_rec.full()),  dims=[[2]*N, [2]*N])
rho_true_qobj = qt.Qobj(np.array(rho_true.full()), dims=[[2]*N, [2]*N])
F = qt.fidelity(rho_true_qobj, rho_rec_qobj)**2
print(f"Fidelität: {F:.4f}")

File: research lab/processing.jl
using LinearAlgebra
using Zygote
using Optimisers

include("/home/jesko/Documents/research lab/setup.jl") 

function GenerateRandomDensityMatrixNoZeros(n::Int)
    d = 2^n
    M   = randn(ComplexF64, d, d) + im * randn(ComplexF64, d, d)
    rho = M * M'
    rho /= tr(rho)
    return rho
end



# --------------------------------------------------------------------
# CNOTGate
# --------------------------------------------------------------------
function CNOTGate(n::Int, control::Int, target::Int)
    dim = 2^n
    gate = zeros(ComplexF64, dim, dim)

    for state in 0:(dim-1)
        control_bit = (state >> (n - 1 - control)) & 1
        if control_bit == 1
            # flippe target bit
            new_state = xor(state, 1 << (n - 1 - target))
        else
            new_state = state
        end
        gate[new_state + 1, state + 1] = 1.0
    end

    return gate
end

function randomInitialT(N::Int)
    dim = 2^N
    return randn(dim, dim) + im * randn(dim, dim)
end

function densityMatrixFromT(T::Matrix{ComplexF64})
    AtA = T' * T
    return AtA / tr(AtA)
end

# --------------------------------------------------------------------
# ParseCircuitToMatrix
# --------------------------------------------------------------------
function ParseCircuitToMatrix(text_circuits::Vector{String}, n::Int)
    unitary_list = Matrix{ComplexF64}[]

    for circuit in text_circuits
        U = Matrix{ComplexF64}(I, 2^n, 2^n)

        # Operationen extrahieren
        operations = split(circuit, ")")
        operations = [strip(op, ['(', ')']) for op in operations if !isempty(strip(op))]

        for op in operations
            parts     = split(op, ":")
            length(parts) < 2 && continue
            gate_name = String(parts[1])
            indices   = parse.(Int, split(parts[2], ","))

            if gate_name == "RX90"
                rx = [cos(π/4) -im*sin(π/4); -im*sin(π/4) cos(π/4)]
                gate = Matrix{ComplexF64}(I, 2^n, 2^n)
                # Tensorprodukt
                full = ones(ComplexF64, 1, 1)
                for q in 0:(n-1)
                    full = kron(full, q == indices[1] ? rx : Matrix{ComplexF64}(I, 2, 2))
                end
                U = full * U

            elseif gate_name == "RY90"
                ry = [cos(π/4) -sin(π/4); sin(π/4) cos(π/4)]
                full = ones(ComplexF64, 1, 1)
                for q in 0:(n-1)
                    full = kron(full, q == indices[1] ? ry : Matrix{ComplexF64}(I, 2, 2))
                end
                U = full * U

            elseif gate_name == "CNOT"
                U = CNOTGate(n, indices[1], indices[2]) * U

            elseif gate_name == "H"
                h = [1 1; 1 -1] / sqrt(2)
                full = ones(ComplexF64, 1, 1)
                for q in 0:(n-1)
                    full = kron(full, q == indices[1] ? h : Matrix{ComplexF64}(I, 2, 2))
                end
                U = full * U
            else
                error("Unbekanntes Gate: $gate_name")
            end
        end

        push!(unitary_list, U)
    end

    return unitary_list
end

# --------------------------------------------------------------------
# PrepareState
# --------------------------------------------------------------------
function PrepareState(initial_text::String, N::Int)
    U_list = ParseCircuitToMatrix([initial_text], N)
    U      = U_list[1]

    # |00...0⟩
    psi0      = zeros(ComplexF64, 2^N)
    psi0[1]   = 1.0

    psi       = U * psi0
    rho_ideal = psi * psi'
    return rho_ideal
end

# --------------------------------------------------------------------
# FlattenList
# --------------------------------------------------------------------
# Richtig:
function FlattenList(nested::Vector{Vector{T}}) where {T}
    return vcat(nested...)
end
# --------------------------------------------------------------------
# BlockSeparator
# --------------------------------------------------------------------
function BlockSeparator(rho::Matrix{ComplexF64}, selective_blocks::Vector{Int}, N::Int)
    @assert all(0 .<= selective_blocks .< 2^N) "Alle Blöcke müssen < 2^N sein"

    result = zeros(ComplexF64, 2^N, 2^N)

    for block in selective_blocks
        block_bits = digits(block, base=2, pad=N) |> reverse

        for i in 0:(2^N - 1)
            i_bits = digits(i, base=2, pad=N) |> reverse
            for j in 0:(2^N - 1)
                j_bits = digits(j, base=2, pad=N) |> reverse

                # Prüfe ob (i,j) zu diesem Block gehört
                matches = all(
                    block_bits[k] == 1 ? i_bits[k] != j_bits[k] : i_bits[k] == j_bits[k]
                    for k in 1:N
                )

                if matches
                    result[i+1, j+1] += rho[i+1, j+1]
                end
            end
        end
    end

    return result
end

# --------------------------------------------------------------------
# DataPredictFromRho
# --------------------------------------------------------------------
function DataPredictFromRho(rho::Matrix{ComplexF64}, unitaries::Vector{Matrix{ComplexF64}}, shots::Int)
    predictions = Vector{Vector{Float64}}()

    for U in unitaries
        # pᵢ = diag(U · ρ · U†) · shots
        probs = real(diag(U * rho * U'))
        push!(predictions, probs .* shots)
    end

    return predictions
end

# --------------------------------------------------------------------
# DataPredictFromRhoSampled
# --------------------------------------------------------------------
function DataPredictFromRhoSampled(rho::Matrix{ComplexF64}, unitaries::Vector{Matrix{ComplexF64}}, shots::Int)
    samples = Vector{Vector{Int}}()

    for U in unitaries
        probs = real(diag(U * rho * U'))
        probs = max.(probs, 0)
        probs = probs / sum(probs)

        # Multinomial sampling
        counts = zeros(Int, length(probs))
        for _ in 1:shots
            r   = rand()
            cum = 0.0
            for (k, p) in enumerate(probs)
                cum += p
                if r <= cum
                    counts[k] += 1
                    break
                end
            end
        end

        push!(samples, counts)
    end

    return samples
end

# --------------------------------------------------------------------
# ProcessData
# --------------------------------------------------------------------
# Fix: kein conj
function ProcessData(
    data::Vector{Vector{Int}},
    unitaries::Vector{Matrix{ComplexF64}},
    selective_blocks::Vector{Int},
    shots::Int,
    N::Int;
    rho_ideal::Union{Matrix{ComplexF64}, Nothing} = nothing,
    lr::Float64      = 0.1,
    decay::Float64   = 0.999,
    iterations::Int  = 600,
    patience::Int    = 50,
    tol::Float64     = 1e-10
)
    measured_flat = Float64.(vcat(data...))

    function loss(T)
        rho       = T' * T
        rho       = rho / tr(rho)
        predicted = vcat([real(diag(U * rho * U')) for U in unitaries]...)
        return -sum(measured_flat .* log.(predicted .+ 1e-8)) / shots
    end

    T = randomInitialT(N)

    L0, _ = Zygote.withgradient(loss, T)
    println("=== DEBUG Julia ===")
    println("Loss initial: ", L0)

    m  = zeros(ComplexF64, size(T))
    v  = zeros(Float64,    size(T))
    β₁, β₂, ε_adam = 0.9, 0.999, 1e-8
    losses     = Float64[]
    no_improve = 0

    for i in 1:iterations
        L, grads = Zygote.withgradient(loss, T)
        g = grads[1]  # ← kein conj!

        η  = lr * decay^i
        m  = β₁ .* m + (1 - β₁) .* g
        v  = β₂ .* v + (1 - β₂) .* abs2.(g)
        m̂  = m ./ (1 - β₁^i)
        v̂  = v ./ (1 - β₂^i)
        T  = T - η .* m̂ ./ (sqrt.(v̂) .+ ε_adam)

        push!(losses, L)

        if i % 100 == 0
            println("Iteration $i | Loss: $(round(L, digits=6))")
        end

        if length(losses) > 1
            if abs(losses[end] - losses[end-1]) < tol
                no_improve += 1
            else
                no_improve = 0
            end
        end
        no_improve >= patience && break
    end

    rho_rec = T' * T
    rho_rec = rho_rec / tr(rho_rec)

    return BlockSeparator(rho_rec, selective_blocks, N)
end
N = 4

println("── ProcessData N=4 ──")
rho_true = GenerateRandomDensityMatrixNoZeros(N)

# Automatisch generieren
blocks    = collect(0:(2^N-1))
ent_circs = BuildEntanglerBlocks(blocks, N)
circuits  = FlattenList(ent_circs)
println("Anzahl Schaltkreise: ", length(circuits))

Us_all = ParseCircuitToMatrix(circuits, N)
shots  = 1000 * 2^N
data   = DataPredictFromRhoSampled(rho_true, Us_all, shots)

rho_rec = ProcessData(
    data, Us_all, blocks, shots, N;
    iterations    = 3000,
    lr            = 0.1,
    decay         = 0.9999,  # ← langsamer sinkend
    patience      = 200,     # ← länger warten
    tol           = 1e-12    # ← strenger
)

F = real(tr(sqrt(sqrt(rho_true) * rho_rec * sqrt(rho_true))))^2
println("Fidelität: ", round(F, digits=4))